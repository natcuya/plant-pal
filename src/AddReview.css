
.AddReview h2{
    color: green;
}
.AddReview{
    background-color: rgba(255, 255, 255, 0.87);
    color: green;
    font-size: 30px;
}

label{
  margin:30px;
}
#review-content.review-input{
    margin: 10px;
    height: 80px;
    width: 300px;
}
.add-review-button{
    margin: 30px;
    height: 50px;
    width: 100px;
    color:green;
}
#rating.rating{
    height: 30px;
    width: 60px;
    margin:-15px;
    padding:10px;
}