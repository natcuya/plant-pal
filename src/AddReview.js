import React from 'react';
import ValidationError from './ValidationError';
import ApiContext from './ApiContext';
import config from './config';
//import './AddReview.css';

class AddReview extends React.Component {
    static contextType = ApiContext
    constructor(props) {
        super(props)
        this.state = {
            rating: {
                value: '',
                touched: false
            },
           content: {
                value: '',
                touched: false
            },
            plantid: {
                value: '',
                touched: false
            }
        }
    }

  updateReviewRating(rating) {
    this.setState({rating: {value:rating, touched: true}})
  }

  updateReviewContent(content) {
    this.setState({content: {value:content, touched: true}})
  }

  updateReviewPlant(plantid) {
    this.setState({plantid: {value: plantid, touched: true}})
  }

  handleSubmit(event) {
    event.preventDefault();
    const { rating, content, plantid } = this.state
    const reviewToAdd = {
      rating: rating.value,
      content: content.value,
      plantid: Number(plantid.value),
    }
    fetch(`${config.API_ENDPOINT}/reviews`, {
      method: 'POST',
      headers: {
        'content-type': 'application/json'
      },
      body: JSON.stringify(reviewToAdd)
    })
    .then(res => {
      console.log(res.body)
      if(!res.ok)
        return res.json().then(e=>Promise.reject(e))
      return res.json()
    })
    .then(review  => {
      console.log('right before addreview pushes to landing')
      this.context.addReview(review)
      this.props.history.push(`/plants/${review.plantid}`)
    //  this.props.history.push('/')
    })
    .catch(error => {
      console.error(error)
    })
  }

  validatePlantName() {
    const name = this.state.name.value.trim();
    if(name.length === 0) {
      return 'Plant name is required'
    }
  }

  validateReviewRating() {
    const rating = this.state.rating.value.trim();
    if(rating.length === 0) {
      return 'Rating is required'
    }
  }

  validateContentName() {
    const content = this.state.content.value.trim();
    if(content.length === 0) {
      return 'Content is required'
    } else if (content.length > 500) {
      return 'Content must be less than 500 characters.'
    }
  }

  render() {
    const ratingError = this.validateReviewRating();
    const contentNameError = this.validateContentName();
    const plantNameError = this.validatePlantName();
    return (
      <section>
        <form className="add-review" onSubmit={ e=> this.handleSubmit(e)}>
          <h2>Create Review</h2>
          <div className="form-group">
            <label htmlFor="review-rating">
              
            </label>
            <input type="text" className="review-input" name="review-rating" id="review-rating" onChange={e => this.updateReviewRating(e.target.value)}>
            </input>
            {this.state.rating.touched && <ValidationError message={ratingError} />}
          </div>
          <div className="form-group">
            <label htmlFor="content-name">
              Content
            </label>
            <textarea className="content-input" name="content-name" id="content-name" onChange={e=> this.updateReviewContent(e.target.value)}/>
            {this.state.content.touched && <ValidationError message={contentNameError} />}
          </div>
          <div className="form-group">
            <label htmlFor="plant-select">Choose a plant:</label>
            <select name="plant-select" id="plant-select" onChange={e=>this.updateReviewPlant(e.target.value)}>
              <option value="">Please select a plant</option>
              {this.context.plants.map(plant => {
                return (
                  <option key={plant.id} value={plant.id}>{plant.name}</option>
                )
              })}
            </select>
            {this.state.plant.touched && <ValidationError message={plantNameError} />}
          </div>
          <div>
          <select
            required
            aria-label='Rate this plant!'
            name='rating'
            id='rating'
          >
            <option value='1'>1 Star</option>
            <option value='2'>2 Stars</option>
            <option value='3'>3 Stars</option>
            <option value='4'>4 Stars</option>
            <option value='5'>5 Stars</option>
          </select>
        </div>
          <button className="add-review-button" disabled={this.validateReviewName() || 
            this.validateContentName() || this.validatePlantName()
            }
           >
            Add Review</button>
        </form>
      </section>
    )
  }
}

export default AddReview