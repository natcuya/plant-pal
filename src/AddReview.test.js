  import React from 'react';
import ReactDOM from 'react-dom';
import AddReview from './AddReview';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<AddReview />, div);
  ReactDOM.unmountComponentAtNode(div);
});
it('renders the UI as expected', () => {
    const tree = renderer
      .create(<AddReview name="AddReview"/>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
    });