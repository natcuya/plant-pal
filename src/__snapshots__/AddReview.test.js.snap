// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the UI as expected 1`] = `
<section
  className="AddReview"
>
  <form
    className="add-review"
    onSubmit={[Function]}
  >
    <h2>
      Create Review
    </h2>
    <div
      className="form-group"
    >
      <label
        htmlFor="review-content"
      >
        Write your review:
      </label>
      <br />
      <input
        className="review-input"
        id="review-content"
        name="review-content"
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="plant-select"
      >
        Choose a plant:
      </label>
      <br />
      <select
        id="plant-select"
        name="plant-select"
        onChange={[Function]}
      >
        <option
          value=""
        >
          Please select a plant
        </option>
      </select>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="rating"
      >
        Rate this plant (1-5): 
      </label>
      <input
        className="rating"
        id="rating"
        max="5"
        min="1"
        name="rating"
        onChange={[Function]}
        type="number"
      />
    </div>
    <button
      className="add-review-button"
      disabled="Content is required"
    >
      Add Review
    </button>
  </form>
</section>
`;
